Make endpoint routing allocation free in common scenarios

davidfowl opened the pull request:
- This change tries to remove the EndpointSelectorContext allocation by making it a wrapper struct over the HttpContext. Unlike before, the HttpContext gets mutated once any component in the routnig pipeline sets a non null endpoint. This used to happen after the processing was complete.
- This change also implements the IRouteValuesFeature and IEndpointFeature in HttpProtocol to avoid the feature allocation and feature collection version churn.

Open issues:
- The log that checks if the endpoit is non null is wrong. If code sets an end point manually before routing runs it'll say match success. We'd need to compare the end point before and after routing to see if it changed to be accurate.
- Currently I removed the IRoutingFeature (partially why this is a draft). We need to push it into the HttpRequest feature in order to make it lazy. We should only be allocating this thing if it's requested. I know @rynowak wants it to go away but maybe it can't right now?
- I didn't do thourough testing yet, it's late and I'll just wait for the CI to explode.

Plaintext without logging now indistinguishable from middleware from an allocation profile:

![image](https://user-images.githubusercontent.com/95136/56845708-1dfe1380-687a-11e9-9f5c-edc87371153f.png)

Fixes https://github.com/aspnet/AspNetCore/issues/6596

davidfowl committed:
Make endpoint routing allocation free in common scenarios
    - This change tries to remove the EndpointSelectoContext allocation by making it a wrapper struct over the HttpContext. Unlike before, the HttpContext gets mutated once any component in the routnig pipeline sets a non null endpoint. This used to happen after the processing was complete.
    - This change also implements the IRouteValuesFeature and IEndpointFeature in HttpProtocol to avoid the feature allocation and feature collection version churn.

davidfowl commented on src/Http/Routing/test/UnitTests/EndpointSelectorContextTest.cs:
```diff
@@ -9,51 +9,52 @@
-            // Assert
-            Assert.Empty(routeData.DataTokens);
-        }
+        //[Fact]
```
TBD

rynowak replied:
hue

rynowak replied:
I'd imagine the best thing to do would be to make this part of `ActionContext` - and basically say that MVC polyfills this for ER and nothing else.

rynowak replied:
I can send a commit if you want to see what that looks like.

davidfowl replied:
Make a change and I can cherry pick it (like I did with your other change).

davidfowl committed:
Code check

davidfowl commented on src/Http/Routing/src/EndpointRoutingMiddleware.cs:
```diff
@@ -97,8 +97,6 @@ private Task SetRoutingAndContinue(HttpContext httpContext, EndpointSelectorCont
                 // existing state for related features unless we did something.
-                SetFeatures(httpContext, feature);
-
                 Log.MatchSuccess(_logger, feature);
```
This isn't right

rynowak replied:
Well, it used to be lel

rynowak replied:
Yah, to keep the same behaviour you'd have to save and restore endpoint and rvd. 

OR 

we could dicuss this more and see what the right behaviour is when you ahve multiple routes. I know Jimmy is particularly passionate about this topic, but I made us just do something simple.

davidfowl committed:
Fix MVC errors

JamesNK commented on src/Http/Routing/src/EndpointSelectorContext.cs:
```diff
@@ -3,68 +3,52 @@
-                _routeData = null;
+                if (value != null)
+                {
+                    _httpContext.SetEndpoint(value);
```
Ah never mind, you've added it to HttpProtocol.

JamesNK commented on src/Http/Routing/src/EndpointSelectorContext.cs:
```diff
@@ -3,68 +3,52 @@
 namespace Microsoft.AspNetCore.Routing
 {
-    public sealed class EndpointSelectorContext : IEndpointFeature, IRouteValuesFeature, IRoutingFeature
+    public struct EndpointSelectorContext
```
This will break anyone who has written and distributed an endpoint selector. Can you think of any other side effects @rynowak?

davidfowl replied:
I’m like 87% sure we’re ok breaking that.

rynowak replied:
I think we should break it. There are a few other things we should think about breaking.

rynowak replied:
I think that set of people is probably just @commonsensesoftware - either way we'll announcement it.

davidfowl replied:
Marked the PR as breaking change

commonsensesoftware replied:
Thanks for the heads up. I looked at this change and it looks to be a non-issue for me. Best as I can tell, this will be 100% source compatible with what I already have. The only write scenario I have is explicitly setting the endpoint as a short-circuit for invalid client requests (ex: duplicate API versions, no API version, etc). It appears this change doesn't alter that behavior.

What will this mean for route constraints - anything? I believe today they are still called in _Endpoint Routing_, but supply `null` for the **IRouter** parameter. That doesn't affect me, but I was wondering if I should expect a possible change there as well. I only use a route constraint for URL segment versioning so I don't need to perform _magic string_ parsing on the URL to find where the version value is in the path.

rynowak replied:
There's no change for `IRouteContraint`, they are used by both routing systems.

rynowak commented:
> Currently I removed the IRoutingFeature (partially why this is a draft). We need to push it into the HttpRequest feature in order to make it lazy. We should only be allocating this thing if it's requested. I know @rynowak wants it to go away but maybe it can't right now?

I think we should lean into this more. MVC could produce a `RouteData` eagerly or on-demand.

rynowak commented on src/Http/Routing/perf/Matching/MatcherSingleEntryBenchmark.cs:
```diff
@@ -68,12 +64,11 @@ public async Task Dfa()
-            // This is required to make the legacy router implementation work with global routing.
-            httpContext.Features.Set<IEndpointFeature>(feature);
+            //// This is required to make the legacy router implementation work with global routing.
+            //httpContext.Features.Set<IEndpointFeature>(feature);
```
I think this can just be removed right? Now this feature is always present.

davidfowl replied:
Yea I wasn’t sure at first

rynowak replied:
It's ok. we all feel that way sometimes.

davidfowl commented:
> I think we should lean into this more. MVC could produce a RouteData eagerly or on-demand. I want to understand what would happen if we did something other than implementing it on the context directly.

What do you want to do?

rynowak commented on src/Http/Routing/src/EndpointRoutingMiddleware.cs:
```diff
@@ -109,15 +107,6 @@ private Task SetRoutingAndContinue(HttpContext httpContext, EndpointSelectorCont
-            // IRouteValuesFeature and IRoutingFeature
-            httpContext.Features.Set<IRoutingFeature>(context);
-            httpContext.Features.Set<IRouteValuesFeature>(context);
-            httpContext.Features.Set<IEndpointFeature>(context);
```
I'm really excited to see this disappear. I expect this to be a huge perf win.

rynowak commented on src/Http/Routing/src/EndpointSelectorContext.cs:
```diff
@@ -3,68 +3,52 @@
-        public Endpoint Endpoint { get; set; }
+        public EndpointSelectorContext(HttpContext httpContext)
+        {
+            _httpContext = httpContext;
```
nullified check

rynowak commented on src/Http/Routing/src/EndpointSelectorContext.cs:
```diff
@@ -3,68 +3,52 @@
-
-                // RouteData will be created next get with new Values
-                _routeData = null;
+                if (value != null)
```
This should still mutate the context. It needs to be possible to set it to null.

rynowak commented on src/Http/Routing/src/EndpointSelectorContext.cs:
```diff
@@ -3,68 +3,52 @@
-                    }
-                }
-
-                return _routeData;
```
yeah, looking at this code, I think I'd feel really good about making `RouteData` lazy-init in MVC.

rynowak commented on src/Http/Routing/test/UnitTests/Matching/TreeRouterMatcherBuilder.cs:
```diff
@@ -99,11 +99,7 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
-                var context = (EndpointSelectorContext)routeContext.HttpContext.Features.Get<IEndpointFeature>();
-
-                // This is needed due to a quirk of our tests - they reuse the endpoint feature.
-                context.Endpoint = null;
```
Still needed eh?

davidfowl replied:
Is it? I couldn't figure that out.

davidfowl replied:
Yep

rynowak commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs:
```diff
@@ -18,16 +19,18 @@
+                                          IHttpMaxRequestBodySizeFeature,
+                                          IHttpResponseStartFeature,
+                                          IEndpointFeature,
+                                          IRouteValuesFeature
```
<3

rynowak commented:
I'm really surprised and pleased how streamlined this is.

rynowak commented:
> 
> 
> > I think we should lean into this more. MVC could produce a RouteData eagerly or on-demand. I want to understand what would happen if we did something other than implementing it on the context directly.
> 
> What do you want to do?

I put a suggestion elsewhere while I was reading through the PR - but I think it's clearer to me now.

I would have MVC create the `RouteData` and then pass it in here: https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.Abstractions/src/ActionContext.cs#L64

This will result in an allocation for now. But we have two paths forward:
- Try to make RouteData lazy-init inside MVC (might be hard)
- Wait until MVC removes support for `IRouter` to remove it. 

We would announce/document that Endpoint Routing doesn't use `IRouteDataFeature`. If we have extension methods that interact with that, we should obsolete them now or in 4.0.

rynowak commented:
Here's a really stupid question. Given that this is just a wrapper for HttpContext, why does it need to exist at all. Since we're mutating HttpContext in place I think there's just no purpose to it.

JamesNK commented:
I does provide clean API for people to set routing stuff onto the context. For example, SetEndpoint is an extension method that lives in its own namespace. It isn't easily discoverable.

rynowak commented:
> I does provide clean API for people to set routing stuff onto the context. For example, SetEndpoint is an extension method that lives in its own namespace. It isn't easily discoverable. 

We could change that tho?

Commit:
Make MVC polyfill RouteData

Since we're not implementing IRoutingFeature in endpoint routing anymore
we can just create the RouteData inside MVC.

rynowak commented:
@davidfowl 4e4538407a05de27d6c6a9c58a41b5ec0747ac1d

davidfowl commented:
> I does provide clean API for people to set routing stuff onto the context. For example, SetEndpoint is an extension method that lives in its own namespace. It isn't easily discoverable.

I was considering lifting Endpoint to `HttpContext`, I wasn't sure though. Feels like it might be crossing the generic HttpContext line (maybe I'm just paranoid).

davidfowl commented:
> Here's a really stupid question. Given that this is just a wrapper for HttpContext, why does it need to exist at all. Since we're mutating HttpContext in place I think there's just no purpose to it.

The more I look at this change the more I feel like we should just pass the HttpContext instead of having the EndpointSelectorContext.

rynowak commented:
> I does provide clean API for people to set routing stuff onto the context. For example, SetEndpoint is an extension method that lives in its own namespace. It isn't easily discoverable.

I should have made this clear, but I guess I'm not optimizing for it being easy. It's somewhat rare to set these properties in a policy, it's for the short-circuit case.

rynowak commented on src/Http/Routing/src/Matching/DfaMatcher.cs:
```diff
@@ -34,11 +34,6 @@ public sealed override Task MatchAsync(HttpContext httpContext, EndpointSelector
-            {
-                throw new ArgumentNullException(nameof(context));
-            }
-
```
Take a look here: https://github.com/aspnet/AspNetCore/pull/9774/files#diff-1473cfa07c9b9a3090e60b371891a1f4L310 This has to do something different now to process short-circuits.

davidfowl committed:
Fixed some things
- Fixed short circuiting
- Fixed logs

davidfowl committed:
Remove reliance on EndpointSelectorContext in unrelated tests

davidfowl committed:
Restore unit test quirk

davidfowl committed:
Reset endpoint and RouteValues in Reset

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:
```diff
@@ -326,6 +329,7 @@ public void Reset()
         {
             _onStarting?.Clear();
             _onCompleted?.Clear();
+            _routeValues?.Clear();
```
@rynowak and @JamesNK this opens the door to re-using RouteValues but we'd need to make changes to routing to avoid reassiging this field and instead just adding values. It doesn't need to happen in this PR though but give it some thought.

rynowak replied:
Reusing route values will involves some surgery to replace/reuse the underlying array. The 'slots' optimization that's use to build the route values was a huge perf win, and that likely is more impactful is saving an alloc of the wrapper object. 

Put another way, building the dictionary using `Add` and other methods that set values in the dictionary is slow. Precomputing which indexes the values belong in is fast. So if we want to look at reusing RVD, we also have to look at manipulating its guts directly.

davidfowl replied:
It also means that CandidateState needs to expose the slots rather than the RVD https://github.com/aspnet/AspNetCore/blob/2b2aeee9f869ec43f65c188a66d9285858823628/src/Http/Routing/src/Matching/CandidateState.cs#L53. That complicates the API a bit.

JamesNK replied:
RVD has a constructor to wrap the underlying array. RVD could be extended to provide a way to get the underlying array instance.

davidfowl replied:
It would only need to allocate if there were complex segments or constraints:
 https://github.com/aspnet/AspNetCore/blob/2b2aeee9f869ec43f65c188a66d9285858823628/src/Http/Routing/src/Matching/DfaMatcher.cs#L163

https://github.com/aspnet/AspNetCore/blob/2b2aeee9f869ec43f65c188a66d9285858823628/src/Http/Routing/src/Matching/DfaMatcher.cs#L153

It seems like we need to pass RouteValues to legacy APIs that have them in the signature. Complex segments could be fixed it we cared but it's likely they are uncommon enough to not worry about (this is an internal API).

`IRouteConstraint` takes `RouteValues` as well (and has an `IRouter` in the signature....) so we need to allocate there.

In the other cases, we can make a new API like `FromArray(slots)` but an instance version (it would need to deal with edge cases though). Anyhow, thats future work.

davidfowl commented on src/Http/Routing/test/testassets/RoutingWebSite/UseEndpointRoutingStartup.cs:
```diff
@@ -173,7 +173,7 @@ private void SetupBranch(IApplicationBuilder app, string name)
             app.UseEndpoints(endpoints =>
             {
-                endpoints.MapGet("api/get/{id}", (context) => context.Response.WriteAsync($"{name} - API Get {context.GetRouteData().Values["id"]}"));
+                endpoints.MapGet("api/get/{id}", (context) => context.Response.WriteAsync($"{name} - API Get {context.Request.RouteValues["id"]}"));
```
@rynowak RouteData is null now.

JamesNK replied:
It should be created lazily if GetRouteData is called for backwards compat

davidfowl replied:
Maybe, I added a change that news it up over the current route values. We'll need to decide what to do here all up.

davidfowl replied:
This is fixed but I'm going to leave the 3.0 way in here

davidfowl committed:
Fixed some tests
- Make GetRouteValue fallback to the IRouteValues feature

davidfowl commented on src/Http/Routing.Abstractions/src/RoutingHttpContextExtensions.cs:
```diff
@@ -46,8 +47,8 @@ public static object GetRouteValue(this HttpContext httpContext, string key)
 
-            var routingFeature = httpContext.Features[typeof(IRoutingFeature)] as IRoutingFeature;
-            return routingFeature?.RouteData.Values[key];
+            return httpContext.Features.Get<IRoutingFeature>()?.RouteData.Values[key] ??
```
@rynowak do we want this to work? Is this a legacy method?

JamesNK replied:
GetRouteValue/GetRouteData are legacy

davidfowl replied:
Does that mean GetRouteValue (which is visible by default) doesn't work when endpoint routing runs?

JamesNK replied:
It worked because the RouteData was created when requested and wrapped the route values - https://github.com/aspnet/AspNetCore/blob/master/src/Http/Routing/src/EndpointSelectorContext.cs#L43-L68

GetRouteValue should be changed to get the value from httpContext.Request.RouteValues to avoid allocating a RouteData.

davidfowl replied:
OK I made some changes.  The last thing to figure out is data tokens. I have some thoughts.

davidfowl commented on src/Http/Routing/test/UnitTests/EndpointRoutingMiddlewareTest.cs:
```diff
@@ -61,52 +61,6 @@ public async Task Invoke_OnCall_WritesToConfiguredLogger()
         }
 
-        [Fact]
-        public async Task Invoke_BackCompatGetRouteValue_ValueUsedFromEndpointFeature()
```
These were testing interop between RouteData on the IRoutingFeature and IRouteValuesFeature. These might still be valid but we need to make a decision about RouteData first.

JamesNK commented on src/Http/Routing/src/EndpointSelectorContext.cs:
```diff
@@ -3,68 +3,49 @@
 namespace Microsoft.AspNetCore.Routing
 {
-    public sealed class EndpointSelectorContext : IEndpointFeature, IRouteValuesFeature, IRoutingFeature
+    public struct EndpointSelectorContext
```
readonly 😄

davidfowl replied:
The properties are settable.

davidfowl replied:
Hmm, seems to not complain.

JamesNK replied:
It's the struct fields that matter. The properties are modifying HttpContext and friends.

davidfowl replied:
Yep

JamesNK commented on src/Http/Routing/src/EndpointSelectorContext.cs:
```diff
@@ -3,68 +3,49 @@
     {
-        private RouteData _routeData;
-        private RouteValueDictionary _routeValues;
+        private HttpContext _httpContext;
```
readonly 😄

JamesNK commented on src/Http/Routing/test/UnitTests/EndpointSelectorContextTest.cs:
```diff
@@ -1,9 +1,8 @@
-using System.Linq;
 using Microsoft.AspNetCore.Http;
-using Microsoft.AspNetCore.Routing.Matching;
+using Microsoft.AspNetCore.Http.Endpoints;
```
This Get/SetEndpoint + this namespace are new in 3.0. Not a breaking change if we want to move them

davidfowl replied:
Not as part of this PR, that'll affect alot more code than I want to touch

davidfowl committed:
Fixed Kestrel tests

davidfowl committed:
Made GetRouteData work (sorta)

davidfowl committed:
Code check,,,,

davidfowl committed:
Implement GetRouteValue in terms of IRouteValuesFeature
- Set the current route values when using IRouter

davidfowl commented:
@rynowak so GetRouteValue will continue to work forever. It's now based on the new `IRouteValuesFeature`. I updated the old routing middleware to set the RouteValues on the request in the middleware (tiny change). `GetRouteData` will either get the `IRoutingFeature` or it will new up one on the fly over the `HttpRequest.RouteValues`.

DataTokens is the last feature we need to figure out. After looking at the code a bit, help me understand why we need to preserve that feature in the endpoint routing MVC case? Is it just to preserve some back compat?

jo-ninja commented on src/Http/Routing/perf/EndpointRoutingBenchmarkBase.cs:
```diff
@@ -131,12 +129,11 @@ protected RouteEndpoint CreateEndpoint(string template, string httpMethod)
             object ambientValues = null)
         {
-            var feature = new EndpointSelectorContext { RouteValues = new RouteValueDictionary(ambientValues) };
+            
```
nit: Remove blank line.
```suggestion
```

JamesNK commented on src/Http/Routing/src/EndpointRoutingMiddleware.cs:
```diff
@@ -56,68 +56,75 @@ internal sealed class EndpointRoutingMiddleware
 
-                Log.MatchSuccess(_logger, feature);
+            // If there was no mutation of the endpoint then log failure
+            if (ReferenceEquals(feature.Endpoint, previousEndpoint))
```
I think this is wrong. The previous endpoint could be the same as the match endpoint, e.g. someone could:
1. Create a RoutingEndpoint in `UseEndpoints`
2. In middleware prior to routing they could fetch it from the endpoint datasource and set it on the feature
3. Routing middleware matches the endpoint already on feature

feature.Endpoint is set to null at the start of routing. If it is null at the end then there was no match. Check for null here and then restore the previousEndpoint inside the if.

JamesNK replied:
Add a unit test for this

davidfowl replied:
Sounds like an extreme edge case but so that I’m clear, pick the exact same instance that would have matched the endpoint and observe that the logs are bad yes?

JamesNK commented on src/Http/Routing/src/EndpointRoutingMiddleware.cs:
```diff
@@ -56,68 +56,75 @@ internal sealed class EndpointRoutingMiddleware
+
+            // We need to set the endpoint to null here so that we can observe any mutations
+            // made by the matcher
+            feature.Endpoint = null;
```
Is there much perf impact with EndpointSelectorContext.Endpoint fetching the IEndpointFeature from HttpContext each time it gets and sets the endpoint?

What do you think of EndpointSelectorContext not wrapping HttpContext, and instead getting and setting the endpoint on the struct? The endpoint would then be set on HttpContext at the end of routing. Although would that require the struct be passed around by ref?

davidfowl replied:
The struct cannot be passed by ref as all of the methods are async, I don’t think it’ll have any real performance impact and if we wanted to solve this, I’d do it by hoisting the Endpoint to an HttpContext property

rynowak replied:
This needs to preserve route values as well.

davidfowl replied:
Does it need to preserve route values if we short circuit here?

rynowak commented on src/Http/Routing.Abstractions/src/RoutingHttpContextExtensions.cs:
```diff
@@ -46,8 +47,7 @@ public static object GetRouteValue(this HttpContext httpContext, string key)
 
-            var routingFeature = httpContext.Features[typeof(IRoutingFeature)] as IRoutingFeature;
-            return routingFeature?.RouteData.Values[key];
+            return httpContext.Features.Get<IRouteValuesFeature>()?.RouteValues[key];
```
This doesn't really work because you expect `RouteData.Values` to be the source of truth. This implementation allows them to be different collections.  

`RouteData` is actually pretty complex - it is used to push and pop state (by MVC) which won't trigger updates to `httpContext.Request.RouteValues`. Trying to do a faithful job here would require some changes inside `RouteData`. 

I'd suggest just leaving this API alone and even marking it obsolete. If you're using old stuff, do `RouteData.Values`. If you're doing new stuff use `httpContext.Request.RouteValues`. 

I think if we're going to do more here than just leave this alone it will need some more design work.

rynowak replied:
Now that I think about this more I'm objecting to it less. This only really causes problems when you use both sets of APIs (new and old) and use the old ones to set state, and the new ones to read it.

davidfowl replied:
Right, I think this API should just “work”. I’d the 3 collections get out of sync it’s be there was mutation involved and maybe we should update those places to also implement this feature to respect the mutation. I know this is all outside of the scope of this PR but I guess that’s why get for touching IRoutingFeature

rynowak commented:
> DataTokens is the last feature we need to figure out. After looking at the code a bit, help me understand why we need to preserve that feature in the endpoint routing MVC case? Is it just to preserve some back compat?

It's not really back compat. It's a feature that people use that we haven't tried to replace. Users expect to be able to attach a property bag to a *route* and then read those values later. I think the solution that I created in that one commit is good enough. 

Outside of MVC's scenarios I'd prefer to just leave `IRoutingFeature` and `DataTokens` alone so they can keep working the way they do today for that user.

rynowak commented:
> The more I look at this change the more I feel like we should just pass the HttpContext instead of having the EndpointSelectorContext.

Did you decide not to do this?

rynowak commented on src/Http/Routing.Abstractions/src/RoutingHttpContextExtensions.cs:
```diff
@@ -23,8 +24,8 @@ public static RouteData GetRouteData(this HttpContext httpContext)
-            var routingFeature = httpContext.Features[typeof(IRoutingFeature)] as IRoutingFeature;
-            return routingFeature?.RouteData;
+            var routingFeature = httpContext.Features.Get<IRoutingFeature>();
+            return routingFeature?.RouteData ?? new RouteData(httpContext.Request.RouteValues);
```
Should this assign `routingFeature.RouteData`?

davidfowl replied:
I dunno feels wrong and side effecty. I don’t want to create a feature here especially since it really isn’t faithful part of the future. Can you think of a downside to not setting the feature?

rynowak replied:
I guess I'm wondering about the case where you have a routing feature but route data is null. This method would return (but not store) a new RouteData object.

davidfowl replied:
> I guess I'm wondering about the case where you have a routing feature but route data is null. This method would return (but not store) a new RouteData object.

When would that happen?

davidfowl replied:
I’ll make the change but I’m not sure when this would happen...

rynowak commented on src/Http/Routing/src/EndpointRoutingMiddleware.cs:
```diff
@@ -56,68 +56,75 @@ internal sealed class EndpointRoutingMiddleware
+                                           HttpContext httpContext,
+                                           EndpointSelectorContext feature,
+                                           Endpoint previousEndpoint,
+                                           Task<Matcher> matcherTask)
```
aligning parameters is awful, just let vs format it the default way

davidfowl replied:
lol sure I prefer alignment but whatevs

rynowak commented on src/Http/Routing/src/EndpointRoutingMiddleware.cs:
```diff
@@ -56,68 +56,75 @@ internal sealed class EndpointRoutingMiddleware
+
+            // Store the previous endpoint before matching starts, this lets us restore whatever endpoint was
+            // set at the end of matching if nothing matched
+            var previousEndpoint = feature.Endpoint;
```
Is it worth a discussion about whether the current behaviour of the middleware is even desirable? For instance if the routing middleware no-op'ed when there was an endpoint does taht make more sense? 

I don't feel like we ever really answered these questions because we didn't have good scenarios in mind either way.

davidfowl replied:
I was actually thinking that was a better behavior.

davidfowl replied:
I'm going to implement this as it simplifies alot

JamesNK replied:
I'll be annoying and ask whether it should be an option

davidfowl replied:
No.

rynowak replied:
You're not being annoying - but it shouldn't be an option. This should work (and be documented) behaving in a single way.

davidfowl replied:
We can make it an option when the masses complain (I mean @sebastienros )

JamesNK replied:
Sure. I'm fine with it not being an option straight away, if ever.

davidfowl commented:
> It's not really back compat. It's a feature that people use that we haven't tried to replace. Users expect to be able to attach a property bag to a route and then read those values later. I think the solution that I created in that one commit is good enough.

> Outside of MVC's scenarios I'd prefer to just leave IRoutingFeature and DataTokens alone so they can keep working the way they do today for that user.

Either we think it’s useful and it should be hoisted to the IRouteValues feature (per route, per request, items bag) or its not and it’s only applicable when using IRouterBased routing or in Mvc.

Btw does MVC (I haven’t looked) have a heavy reliance on RouteData or is it mostly round tripping user state from route definition to action invocation?

davidfowl commented:
> Did you decide not to do this?

Yes I want to get this in with the current shape and iterate. Once we solve these big rocks and then come back to this later.

davidfowl committed:
Noop if an endpoint is already set

JamesNK commented on src/Http/Routing/src/EndpointRoutingMiddleware.cs:
```diff
@@ -58,62 +57,49 @@ public Task Invoke(HttpContext httpContext)
-            // made by the matcher
-            feature.Endpoint = null;
+            // There's already an endpoint, skip maching completely
+            if (feature.Endpoint != null)
```
Should be an option?

davidfowl committed:
Fix nit

rynowak commented on src/Http/Routing/src/EndpointRoutingMiddleware.cs:
```diff
@@ -93,31 +98,19 @@ static async Task AwaitMatch(EndpointRoutingMiddleware middleware, HttpContext h
         {
-            if (feature.Endpoint != null)
+            // If there was no mutation of the endpoint then log failure
+            if (feature.Endpoint is null)
```
ugh. `is null` is so objectionable.

davidfowl replied:
is so good

rynowak commented on src/Http/Routing/src/EndpointRoutingMiddleware.cs:
```diff
@@ -56,7 +55,13 @@ internal sealed class EndpointRoutingMiddleware
+            // There's already an endpoint, skip maching completely
+            if (feature.Endpoint != null)
+            {
+                return _next(httpContext);
```
This seems like this case should be logged.

davidfowl replied:
Done

rynowak commented on src/Http/Routing/src/RouterMiddleware.cs:
```diff
@@ -40,11 +40,16 @@ public async Task Invoke(HttpContext httpContext)
+                    RouteData = context.RouteData
                 };
 
+                // Set the RouteValues on the current request
```
If you want extra #RyanPoints update this comment to say that this is being down to keep `IRoutingFeature` in sync with `IRouteValuesFeature`. If you want even more extra points include ascii at of N'sync

rynowak commented on src/Http/Routing/test/UnitTests/EndpointRoutingMiddlewareTest.cs:
```diff
@@ -36,6 +35,24 @@ public async Task Invoke_OnCall_SetsEndpointFeature()
         }
 
+        [Fact]
+        public async Task Invoke_SkipsRouting_IfEndpointSet()
```
👍

davidfowl committed:
PR nits
- Add logging when routing skipped because an endpoint was set
- Updated comment
